flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteStateTableSnapshot.java

```diff
@@ -147,7 +147,8 @@ public StateKeyGroupWriter getKeyGroupWriter() {
 					localKeySerializer.serialize(element.key, dov);
 					localStateSerializer.serialize(element.state, dov);
 				};
-			StateSnapshotTransformer<S> stateSnapshotTransformer = owningStateTable.metaInfo.getSnapshotTransformer();
+			StateSnapshotTransformer<S> stateSnapshotTransformer = owningStateTable.metaInfo.
+				getStateSnapshotTransformFactory().createForDeserializedState().orElse(null);
```

Comment:
The part below this line could be refactored to not check again if `stateSnapshotTransformer != null`.